### DISCUSSION TITLE
PWM question

### DISCUSSION THREAD

**Discussion by RKirklanH** [micropython/micropython#12110] [m.iMX RT / Teensy 4.x]
I recently got a teensy 4.0 and 4.1 to play with, and when testing out the latest micropython firmware (as of July 28th, 2023, both the standard and nightly releases), the PWM on both seems to be acting weird. Whenever I use PWM on pins 0 or 1, everything is normal until I give a duty cycle of 0 or 1:
`.duty_u16(0)
.duty_u16(1)`
Whenever I do this for both the teensy 4.0 and 4.1 on pins 1 or 0, instead of setting the duty cycle as low as it should, it instead seems to jump all the way to the highest possible value. I discovered this while fading an LED with PWM, and I saw it repeatedly flicker right as it got to its dimmest. Sure enough, whenever I manually set the duty cycle to 0 or 1 via the REPL, it turns the LED all the way on. When I manually set the duty cycle to 2, however, the LED then becomes as dim as you would expect, not even "on" enough to see any light. I tested this with other pins, and the issue did not repeat itself. Is it possible this is a firmware issue, or is there something to these pins that I do not know that could make them act like this?

**Comment by robert-hh** [comment]
Thank you for the information. I will look into that and fix it, if it's broken. 
Edit: The same should happen at Pins 24 and 25. All four use the same somewhat limited PWM mechanism, different from the other pins.

**Comment by robert-hh** [comment]
There is a PR now to fix that behavior. It may take a while until it is merged.

**Comment by RKirklanH** [comment]
Thank you! I'm glad it wasn't just a me problem. Do you know what was causing it?

**Reply by robert-hh** [reply]
Why should it just be your problem? Some times it takes a while until an error pops up. In short: an underflow of a 16 bit unsigned symbol from 0 to 65535. The actual change is  just disabling the output in that situation. See https://github.com/micropython/micropython/pull/12114/files, file pmw_backport.c, line 113-118. But sinceI were at it, I made more changes in that scenery. The other PWM mechanisms had to adapted to the common behavior of handling  the values 0 and 65536 for duty_u16.
